tests/server/tso/tso_proxy_test.go

```diff
@@ -0,0 +1,203 @@
+	serverCancel context.CancelFunc
+	cluster      *tests.TestCluster
+	leader       *tests.TestServer
+	follower     *tests.TestServer
+
+	pdClient     pdpb.PDClient
+	defaultReq   *pdpb.TsoRequest
+	proxyClient  pdpb.PD_TsoClient
+	clientCtx    context.Context
+	clientCancel context.CancelFunc
+}
+
+func TestTSOProxyTestSuite(t *testing.T) {
+	suite.Run(t, new(tsoProxyTestSuite))
+}
+
+func (s *tsoProxyTestSuite) SetupTest() {
+	re := s.Require()
+
+	var err error
+	s.serverCtx, s.serverCancel = context.WithCancel(context.Background())
+	s.cluster, err = tests.NewTestCluster(s.serverCtx, 2)
+	re.NoError(err)
+
+	re.NoError(s.cluster.RunInitialServers())
+	re.NotEmpty(s.cluster.WaitLeader())
+
+	for _, server := range s.cluster.GetServers() {
+		if server.GetConfig().Name != s.cluster.GetLeader() {
+			s.follower = server
+		} else {
+			s.leader = server
+		}
+	}
+
+	s.pdClient = testutil.MustNewGrpcClient(re, s.follower.GetAddr())
+	clusterID := s.leader.GetClusterID()
+	s.defaultReq = &pdpb.TsoRequest{
+		Header:     testutil.NewRequestHeader(clusterID),
+		Count:      1,
+		DcLocation: tso.GlobalDCLocation,
+	}
+
+	s.reCreateProxyClient()
+}
+
+func (s *tsoProxyTestSuite) reCreateProxyClient() {
+	if s.proxyClient != nil {
+		_ = s.proxyClient.CloseSend()
+		s.clientCancel()
+	}
+	s.proxyClient, s.clientCtx, s.clientCancel = s.createClient()
+}
+
+func (s *tsoProxyTestSuite) createClient() (pdpb.PD_TsoClient, context.Context, context.CancelFunc) {
+	ctx, cancel := context.WithCancel(context.Background())
+	ctx = grpcutil.BuildForwardContext(ctx, s.leader.GetAddr())
+	tsoClient, _ := s.pdClient.Tso(ctx)
+	return tsoClient, ctx, cancel
+}
+
+func (s *tsoProxyTestSuite) TearDownTest() {
+	_ = s.proxyClient.CloseSend()
+	s.clientCancel()
+	s.cluster.Destroy()
+	s.serverCancel()
+}
+
+func (s *tsoProxyTestSuite) verifyProxyIsHealthy(re *require.Assertions) {
+	s.verifyProxyIsHealthyWith(re, s.proxyClient)
+}
+
+func (s *tsoProxyTestSuite) verifyProxyIsHealthyWith(re *require.Assertions, client pdpb.PD_TsoClient) {
+	re.NoError(client.Send(s.defaultReq))
+	resp, err := client.Recv()
+	re.NoError(err)
+	re.Equal(s.defaultReq.GetCount(), resp.GetCount())
+	timestamp := resp.GetTimestamp()
+	re.Positive(timestamp.GetPhysical())
+	re.GreaterOrEqual(uint32(timestamp.GetLogical()), s.defaultReq.GetCount())
+}
+
+func (s *tsoProxyTestSuite) assertReceiveError(re *require.Assertions, errStr string) {
+	re.NoError(s.proxyClient.Send(s.defaultReq))
+	_, err := s.proxyClient.Recv()
+	re.Error(err)
+	re.Contains(err.Error(), errStr)
+}
+
+func (s *tsoProxyTestSuite) TestProxyPropagatesLeaderErrorQuickly() {
+	re := s.Require()
+	s.verifyProxyIsHealthy(re)
+
+	// change leader
+	re.NoError(s.cluster.ResignLeader())
+
+	start := time.Now()
+	s.assertReceiveError(re, "pd is not leader of cluster")
+
+	// verify fails faster than 3 second timeout, otherwise the unavailable time will be too long.
+	re.Less(time.Since(start), time.Second)
```

Comment:
As the comment suggests, should this be changed to `re.Less(time.Since(start), 3*time.Second)`?